src/cpp/fastdds/xtypes/dynamic_types/idl_parser/IdlParser.hpp

```diff
@@ -0,0 +1,1725 @@
+#include <pegtl/analyze.hpp>
+
+#include "IdlGrammar.hpp"
+#include "IdlModule.hpp"
+
+// mimic posix pipe APIs
+#ifdef _MSC_VER
+#   pragma push_macro("popen")
+#   define popen _popen
+#   pragma push_macro("pipe")
+#   define pipe _pipe
+#   pragma push_macro("pclose")
+#   define pclose _pclose
+#endif // ifdef _MSC_VER
+
+// define preprocessor strategy
+#ifdef _MSC_VER
+#   define EPROSIMA_PLATFORM_PREPROCESSOR "cl /EP /I."
+#   define EPROSIMA_PLATFORM_PREPROCESSOR_STRATEGY preprocess_strategy::temporary_file
+#   define EPROSIMA_PLATFORM_PREPROCESSOR_INCLUDES "/I"
+#   define EPROSIMA_PLATFORM_PREPROCESSOR_ERRORREDIR " 2>nul"
+#   define EPROSIMA_PLATFORM_PIPE_OPEN_FLAGS "rt"
+#else
+#   define EPROSIMA_PLATFORM_PREPROCESSOR "cpp -H"
+#   define EPROSIMA_PLATFORM_PREPROCESSOR_STRATEGY preprocess_strategy::pipe_stdin
+#   define EPROSIMA_PLATFORM_PREPROCESSOR_INCLUDES "-I"
+#   define EPROSIMA_PLATFORM_PREPROCESSOR_ERRORREDIR " 2>/dev/null"
+#   define EPROSIMA_PLATFORM_PIPE_OPEN_FLAGS "r"
+#endif // ifdef _MSC_VER
+
+namespace eprosima {
+namespace fastdds {
+namespace dds {
+namespace idlparser {
+
+using namespace tao::TAO_PEGTL_NAMESPACE;
+
+class PreprocessorContext
+{
+public:
+
+    // Preprocessors capability to use shared memory (pipes) or stick to file input
+    enum class preprocess_strategy
+    {
+        pipe_stdin,
+        temporary_file
+    };
+
+    bool preprocess = true;
+    std::string preprocessor_exec = EPROSIMA_PLATFORM_PREPROCESSOR;
+    std::string error_redir = EPROSIMA_PLATFORM_PREPROCESSOR_ERRORREDIR;
+    preprocess_strategy strategy = EPROSIMA_PLATFORM_PREPROCESSOR_STRATEGY;
+    std::string include_flag = EPROSIMA_PLATFORM_PREPROCESSOR_INCLUDES;
+    std::vector<std::string> include_paths;
+
+    std::string preprocess_file(
+            const std::string& idl_file) const
+    {
+        std::vector<std::string> includes;
+        std::string args;
+        for (const std::string& inc_path : include_paths)
+        {
+            args += include_flag + inc_path + " ";
+        }
+
+        std::string cmd = preprocessor_exec + " " + args + idl_file + error_redir;
+
+        EPROSIMA_LOG_INFO(IDLPARSER, "Calling preprocessor with command: " << cmd);
+        std::string output = exec(cmd);
+        EPROSIMA_LOG_INFO(IDLPARSER, "Pre-processed IDL: " << output);
+        return output;
+    }
+
+    std::string preprocess_string(
+            const std::string& idl_string) const;
+
+private:
+
+    template<preprocess_strategy e>
+    std::string preprocess_string(
+            const std::string& idl_string) const;
+
+#ifdef _MSC_VER
+    std::pair<std::ofstream, std::string> get_temporary_file() const
+    {
+        // Create temporary filename
+        std::array<char, L_tmpnam> filename_buffer;
+        errno_t err = tmpnam_s(filename_buffer.data(), filename_buffer.size());
+        if (err != 0)
+        {
+            throw std::runtime_error("Failed to generate a temporary filename.");
```

Comment:
We allow to use exceptions internally. But the exception must be capture before return to user code. In case of an error return a nil `traits<DynamicType>::ref_type`.

Reply:
Try-catch is added in create_type_w_document.